# Approach 2 — Pure Black-Box (ONNX only)

We treat the model as a pure black box via the ONNX runtime.
We perform a fine-grained grid search over the 4 mutable features,
querying the model for each combination to find the minimum-cost
perturbation that flips the prediction to approved.

In [1]:
import pandas as pd
import numpy as np
import onnxruntime as rt
from itertools import product
import io
from contextlib import redirect_stdout
import warnings
warnings.filterwarnings("ignore")

## 1. Load data and ONNX model

In [2]:
df = pd.read_csv("rejected_applicants.csv")

sess = rt.InferenceSession("loan_approval_model.onnx")
input_name = sess.get_inputs()[0].name
label_name = sess.get_outputs()[0].name
prob_name = sess.get_outputs()[1].name

feature_names = [
    "Age", "Credit_History_Length", "Requested_Loan_Amount",
    "Declared_Income", "Co_Applicant_Income", "Loan_Term_Months",
    "Past_Bankruptcies",
]

X = df[feature_names].copy()
print(f"Input shape expected: {sess.get_inputs()[0].shape}")
print(f"Features: {feature_names}")

Input shape expected: [None, 7]
Features: ['Age', 'Credit_History_Length', 'Requested_Loan_Amount', 'Declared_Income', 'Co_Applicant_Income', 'Loan_Term_Months', 'Past_Bankruptcies']


## 2. ONNX prediction helpers

In [3]:
def predict_batch(X_array):
    """Run ONNX prediction on a batch. Returns (labels, probs_of_class_1)."""
    X_f32 = X_array.astype(np.float32)
    labels, probs = sess.run([label_name, prob_name], {input_name: X_f32})
    return labels, np.array([p[1] for p in probs])

def predict_single(row_array):
    """Convenience wrapper for a single row."""
    labels, probs = predict_batch(row_array.reshape(1, -1))
    return labels[0], probs[0]

## 3. Identify rejected applicants and rank by closeness to boundary

In [4]:
all_labels, all_probs = predict_batch(X.values)
df["approval_prob"] = all_probs
df["prediction"] = all_labels

rejected = df[df["prediction"] == 0].sort_values("approval_prob", ascending=False)
print(f"Total rejected: {len(rejected)}")
rejected.head(10)

Total rejected: 50


,Applicant_ID,Age,Credit_History_Length,Requested_Loan_Amount,Declared_Income,Co_Applicant_Income,Loan_Term_Months,Past_Bankruptcies,approval_prob,prediction
13,13,46,10,33064,40663,0,12,0,0.245652,0
33,33,67,15,5510,37593,20000,60,1,0.173880,0
48,48,56,3,41913,34858,40000,12,0,0.168014,0
45,45,57,4,34894,68733,60000,36,1,0.139926,0
8,8,52,19,11476,55968,0,36,1,0.127839,0
1,1,31,16,44181,47038,60000,24,1,0.124942,0
2,2,65,10,27058,75310,60000,60,1,0.119347,0
15,15,35,1,48371,40872,20000,12,0,0.115404,0
14,14,36,15,23676,49053,40000,24,1,0.113988,0
35,35,26,2,29496,54067,60000,24,1,0.112139,0


## 4. Cost function

In [5]:
# Load and reuse the exact implementation from Cost.txt
with open("Cost.txt", "r", encoding="utf-8") as f:
    cost_source = f.read()

cost_ns = {}
exec(cost_source, cost_ns)
calculate_attack_cost = cost_ns["calculate_attack_cost"]

## 5. Define the search grid

We build a fine grid for each mutable feature in the allowed direction:
- **Requested_Loan_Amount**: decrease from original down to 0 in steps of \$500
- **Declared_Income**: increase from original up to \$200,000 in steps of \$1,000
- **Co_Applicant_Income**: increase from original up to \$100,000 in steps of \$1,000
- **Loan_Term_Months**: increase from original up to 120 in steps of 1 month

Choosing the step sizes to match the cost granularity ensures we don't miss a
cheaper option (the cost function itself uses \$500 / \$1,000 / 1-month units).

In [6]:
def build_grid(orig_row, caps=None):
    """Build bounded candidate values for each mutable feature."""
    if caps is None:
        caps = {
            "max_loan_decrease": 10000,
            "max_income_increase": 30000,
            "max_co_income_increase": 30000,
            "max_term_increase": 24,
        }

    loan_orig = float(orig_row["Requested_Loan_Amount"])
    income_orig = float(orig_row["Declared_Income"])
    co_income_orig = float(orig_row["Co_Applicant_Income"])
    term_orig = float(orig_row["Loan_Term_Months"])

    # Keep step sizes aligned with the cost function granularity.
    loan_min = max(0.0, loan_orig - caps["max_loan_decrease"])
    loan_candidates = np.arange(loan_min, loan_orig + 1, 500)
    loan_candidates = np.unique(np.append(loan_candidates, loan_orig))

    income_max = income_orig + caps["max_income_increase"]
    income_candidates = np.arange(income_orig, income_max + 1, 1000)
    income_candidates = np.unique(np.append(income_candidates, income_orig))

    co_income_max = co_income_orig + caps["max_co_income_increase"]
    co_income_candidates = np.arange(co_income_orig, co_income_max + 1, 1000)
    co_income_candidates = np.unique(np.append(co_income_candidates, co_income_orig))

    term_max = term_orig + caps["max_term_increase"]
    term_candidates = np.arange(term_orig, term_max + 1, 1)
    term_candidates = np.unique(np.append(term_candidates, term_orig))

    return loan_candidates, income_candidates, co_income_candidates, term_candidates

## 6. Search with cost-ordered pruning

Naively enumerating every combination (potentially millions) is slow with ONNX.
Instead we process in **batches sorted by cost**, so we can stop as soon as
we find the first approved combination — anything later would cost more.

In [7]:
def find_optimal_attack_blackbox(row, feature_names, caps=None, max_grid_points=2_000_000):
    """
    Bounded grid search over mutable features for a single applicant.
    Returns (best_cost, best_modified_row) or (inf, None).
    """
    orig = row[feature_names].values.astype(float)
    orig_series = row[feature_names].copy()

    loan_cands, income_cands, co_cands, term_cands = build_grid(row, caps=caps)
    n_total = len(loan_cands) * len(income_cands) * len(co_cands) * len(term_cands)
    print(f"  Grid sizes — Loan: {len(loan_cands)}, Income: {len(income_cands)}, "
          f"CoIncome: {len(co_cands)}, Term: {len(term_cands)}")
    print(f"  Total grid points: {n_total:,}")

    if n_total > max_grid_points:
        print(f"  Skipping (>{max_grid_points:,} points). Tighten caps for this applicant.")
        return float("inf"), None

    feat_idx_loan = feature_names.index("Requested_Loan_Amount")
    feat_idx_income = feature_names.index("Declared_Income")
    feat_idx_co = feature_names.index("Co_Applicant_Income")
    feat_idx_term = feature_names.index("Loan_Term_Months")

    combo_indices = list(product(
        range(len(loan_cands)),
        range(len(income_cands)),
        range(len(co_cands)),
        range(len(term_cands)),
    ))

    best_cost = float("inf")
    best_row = None

    BATCH = 10_000
    for start in range(0, len(combo_indices), BATCH):
        batch_combo = combo_indices[start:start + BATCH]

        X_batch = np.tile(orig, (len(batch_combo), 1))
        for pos, (i, j, k, l) in enumerate(batch_combo):
            X_batch[pos, feat_idx_loan] = loan_cands[i]
            X_batch[pos, feat_idx_income] = income_cands[j]
            X_batch[pos, feat_idx_co] = co_cands[k]
            X_batch[pos, feat_idx_term] = term_cands[l]

        labels, _ = predict_batch(X_batch)
        approved_in_batch = np.where(labels == 1)[0]

        for pos in approved_in_batch:
            mod_series = pd.Series(X_batch[pos], index=feature_names)
            with redirect_stdout(io.StringIO()):
                c = calculate_attack_cost(pd.DataFrame([orig_series]), pd.DataFrame([mod_series]))
            if c < best_cost:
                best_cost = c
                best_row = X_batch[pos].copy()

    return best_cost, best_row

In [8]:
def find_optimal_attack_cost_capped(row, feature_names, max_cost, caps=None, batch_size=10_000):
    """
    Grid search over all 4 mutable features that only evaluates combinations
    with total cost <= max_cost. Cost is computed from Cost.txt rules BEFORE
    any ONNX query, keeping the search provably complete within the budget.
    """
    if caps is None:
        caps = SEARCH_CAPS_BUDGET

    orig = row[feature_names].values.astype(float)
    orig_series = row[feature_names].copy()

    loan_cands, income_cands, co_cands, term_cands = build_grid(row, caps=caps)

    feat_idx_loan   = feature_names.index("Requested_Loan_Amount")
    feat_idx_income = feature_names.index("Declared_Income")
    feat_idx_co     = feature_names.index("Co_Applicant_Income")
    feat_idx_term   = feature_names.index("Loan_Term_Months")

    loan_orig   = float(row["Requested_Loan_Amount"])
    income_orig = float(row["Declared_Income"])
    co_orig     = float(row["Co_Applicant_Income"])
    term_orig   = float(row["Loan_Term_Months"])

    # Pre-compute per-value cost contributions matching Cost.txt rules exactly
    loan_costs   = np.maximum(0.0, (loan_orig   - loan_cands)   / 500.0)  * 1.0
    income_costs = np.maximum(0.0, (income_cands - income_orig) / 1000.0) * 5.0
    co_costs     = np.maximum(0.0, (co_cands     - co_orig)     / 1000.0) * 3.0
    term_costs   = np.maximum(0.0, (term_cands   - term_orig))            * 2.0

    grid_size = len(loan_cands) * len(income_cands) * len(co_cands) * len(term_cands)
    print(f"  Grid: Loan={len(loan_cands)} Income={len(income_cands)} "
          f"Co={len(co_cands)} Term={len(term_cands)} total={grid_size:,}")

    best_cost = float("inf")
    best_row  = None
    total_checked    = 0
    total_within_cap = 0

    combos_iter = product(
        range(len(loan_cands)),
        range(len(income_cands)),
        range(len(co_cands)),
        range(len(term_cands)),
    )

    exhausted = False
    while not exhausted:
        keep = []

        while len(keep) < batch_size:
            try:
                i, j, k, l = next(combos_iter)
            except StopIteration:
                exhausted = True
                break

            total_checked += 1
            c = loan_costs[i] + income_costs[j] + co_costs[k] + term_costs[l]
            if c <= max_cost:
                keep.append((i, j, k, l))
                total_within_cap += 1

        if not keep:
            continue

        X_batch = np.tile(orig, (len(keep), 1))
        for pos, (i, j, k, l) in enumerate(keep):
            X_batch[pos, feat_idx_loan]   = loan_cands[i]
            X_batch[pos, feat_idx_income] = income_cands[j]
            X_batch[pos, feat_idx_co]     = co_cands[k]
            X_batch[pos, feat_idx_term]   = term_cands[l]

        labels, _ = predict_batch(X_batch)
        approved  = np.where(labels == 1)[0]

        for pos in approved:
            mod_series = pd.Series(X_batch[pos], index=feature_names)
            with redirect_stdout(io.StringIO()):
                exact_cost = calculate_attack_cost(
                    pd.DataFrame([orig_series]), pd.DataFrame([mod_series])
                )
            if exact_cost <= max_cost and exact_cost < best_cost:
                best_cost = exact_cost
                best_row  = X_batch[pos].copy()

    print(f"  Combinations checked: {total_checked:,} | within cap: {total_within_cap:,}")
    return best_cost, best_row


def run_all_rejected_with_cost_cap(rejected_df, max_cost, caps):
    """
    Search every rejected applicant (in probe-sensitivity order) using the
    cost-pruned search over all 4 mutable features.
    No applicant is skipped — probe ranking is used only for ordering.
    """
    capped_results = []

    for rank, (_, row) in enumerate(rejected_df.iterrows()):
        app_id = int(row["Applicant_ID"])
        print(f"\n=== Applicant {app_id} (rank {rank+1}, p={row['approval_prob']:.4f}) ===")

        best_cost, best_row = find_optimal_attack_cost_capped(
            row, feature_names, max_cost=max_cost, caps=caps
        )

        if best_row is None:
            print("  ✗ No flip found under cost cap")
        else:
            print(f"  ✓ Best cost: {best_cost:.2f}")
            capped_results.append((app_id, best_cost, best_row, row))

    capped_results.sort(key=lambda x: x[1])
    return capped_results

## 6. Exploration and Grid Setup

This section defines the search space and helper parameters used by all subsequent experiments.

In [9]:
# Step 1 caps: realistic domain bounds, used for generous-cap baseline attacks
SEARCH_CAPS_FULL = {
    "max_loan_decrease": 50000,       # covers any loan in dataset down to 0
    "max_income_increase": 160000,    # covers incomes up to ~200k
    "max_co_income_increase": 100000, # covers co-income up to 100k
    "max_term_increase": 108,         # covers terms up to 120 months
}

# Step 2 caps: mathematically derived from MAX_COST = 20
# Rationale: baseline attack (below) finds cost = 14; we round up to 20.
# Any combination outside these ranges cannot have cost <= 20, so nothing is missed.
MAX_COST = 20
SEARCH_CAPS_BUDGET = {
    "max_loan_decrease":      int(MAX_COST * 500),           # 10,000
    "max_income_increase":    int((MAX_COST / 5) * 1000),    #  4,000
    "max_co_income_increase": int((MAX_COST / 3) * 1000),    #  6,666
    "max_term_increase":      int(MAX_COST / 2),             #     10
}

MAX_GRID_POINTS = 200_000_000  # generous limit; cost pruning keeps actual queries tiny

print(f"Step 1 caps (full domain, no cost cap):")
for k, v in SEARCH_CAPS_FULL.items():
    print(f"  {k}: {v:,}")

print(f"\nStep 2 caps (derived from MAX_COST={MAX_COST}, provably complete):")
for k, v in SEARCH_CAPS_BUDGET.items():
    print(f"  {k}: {v:,}")

print(f"\nQuick exploration — top rejected applicants:")
display(rejected.head(10)[[
    "Applicant_ID", "approval_prob",
    "Requested_Loan_Amount", "Declared_Income",
    "Co_Applicant_Income", "Loan_Term_Months",
]])

Step 1 caps (full domain, no cost cap):
  max_loan_decrease: 50,000
  max_income_increase: 160,000
  max_co_income_increase: 100,000
  max_term_increase: 108

Step 2 caps (derived from MAX_COST=20, provably complete):
  max_loan_decrease: 10,000
  max_income_increase: 4,000
  max_co_income_increase: 6,666
  max_term_increase: 10

Quick exploration — top rejected applicants:


,Applicant_ID,approval_prob,Requested_Loan_Amount,Declared_Income,Co_Applicant_Income,Loan_Term_Months
13,13,0.245652,33064,40663,0,12
33,33,0.173880,5510,37593,20000,60
48,48,0.168014,41913,34858,40000,12
45,45,0.139926,34894,68733,60000,36
8,8,0.127839,11476,55968,0,36
1,1,0.124942,44181,47038,60000,24
2,2,0.119347,27058,75310,60000,60
15,15,0.115404,48371,40872,20000,12
14,14,0.113988,23676,49053,40000,24
35,35,0.112139,29496,54067,60000,24


## 7. Feature Sensitivity Probing

Before searching, we probe each mutable feature with a **1-cost-unit perturbation** and measure the resulting approval probability gain `Δp`.

Because the model is a RandomForest (piecewise-constant), a small nudge that does not cross any split threshold produces exactly zero gain — even if a larger change on that same feature would flip the prediction. The probe is therefore a **local** signal only and **cannot safely be used to eliminate features from the search**.

What the probe *is* reliable for is **ranking applicants**: a high `total_gain` means the applicant is close to at least one decision boundary, making it cheaper and more likely to flip. We use this order to prioritise the search — most-attackable candidates first.

In [10]:
# One cost-unit perturbation per mutable feature (in the allowed direction)
COST_UNIT_DELTA = {
    "Requested_Loan_Amount": -500.0,   # 1 cost unit = -$500
    "Declared_Income":       +200.0,   # 1 cost unit = +$200  (5 per $1k)
    "Co_Applicant_Income":   +333.33,  # 1 cost unit = +$333  (3 per $1k)
    "Loan_Term_Months":      +0.5,     # 1 cost unit = +0.5 months (2 per month)
}
MUTABLE_FEATURES = list(COST_UNIT_DELTA.keys())
SENSITIVITY_THRESHOLD = 1e-4   # minimum Δp to count a feature as "active"


def probe_applicant(row, feature_names):
    """
    Probe each mutable feature with a 1-cost-unit perturbation.

    Returns
    -------
    p0           : float  - baseline approval probability
    gains        : dict   - {feature: Δp from 1-cost-unit push}
    active       : list   - features with gain > SENSITIVITY_THRESHOLD
    total_gain   : float  - sum of all gains (applicant sensitivity score)
    """
    base = row[feature_names].astype(float).values
    _, p0 = predict_single(base)

    gains = {}
    for fname, delta in COST_UNIT_DELTA.items():
        pert = base.copy()
        idx = feature_names.index(fname)
        new_val = pert[idx] + delta
        # enforce physical lower bounds
        new_val = max(0.0, new_val)
        pert[idx] = new_val
        _, p1 = predict_single(pert)
        gains[fname] = max(0.0, p1 - p0)

    active = [f for f, g in gains.items() if g > SENSITIVITY_THRESHOLD]
    total_gain = sum(gains.values())
    return p0, gains, active, total_gain


# Probe all rejected applicants
probe_rows = []
for _, row in rejected.iterrows():
    p0, gains, active, total_gain = probe_applicant(row, feature_names)
    probe_rows.append({
        "Applicant_ID":   int(row["Applicant_ID"]),
        "p0":             round(p0, 4),
        "gain_Loan":      round(gains["Requested_Loan_Amount"], 5),
        "gain_Income":    round(gains["Declared_Income"], 5),
        "gain_CoIncome":  round(gains["Co_Applicant_Income"], 5),
        "gain_Term":      round(gains["Loan_Term_Months"], 5),
        "total_gain":     round(total_gain, 5),
        "n_active":       len(active),
        "active_features": active,
    })

probe_df = pd.DataFrame(probe_rows).sort_values("total_gain", ascending=False).reset_index(drop=True)

print("Rejected applicants ranked by total sensitivity (sum of per-feature probability gains):")
print(probe_df.drop(columns=["active_features"]).to_string(index=False))
print(f"\nApplicants with at least one active feature: {(probe_df['n_active'] > 0).sum()} / {len(probe_df)}")

Rejected applicants ranked by total sensitivity (sum of per-feature probability gains):
 Applicant_ID     p0  gain_Loan  gain_Income  gain_CoIncome  gain_Term  total_gain  n_active
           33 0.1739    0.16308      0.00000            0.0        0.0     0.16308         1
           28 0.0615    0.02771      0.00000            0.0        0.0     0.02771         1
           13 0.2457    0.02472      0.00000            0.0        0.0     0.02472         1
           20 0.0741    0.02345      0.00000            0.0        0.0     0.02345         1
           36 0.0649    0.01865      0.00000            0.0        0.0     0.01865         1
           48 0.1680    0.01412      0.00452            0.0        0.0     0.01864         2
           22 0.0395    0.00917      0.00000            0.0        0.0     0.00917         1
           10 0.0072    0.00917      0.00000            0.0        0.0     0.00917         1
           47 0.1103    0.00000      0.00886            0.0        0.0     

## 7. Top Probe-Ranked Applicant Attack (Generous Cap)

The probe ranked **Applicant 33** first — it has the highest total sensitivity score (a $500 loan reduction raises its approval probability by ~16 percentage points in one step).

We attack it using a generous cost cap of **50**, which is effectively unconstrained for any practical attack (the best known flip costs 14). Using cost-pruning instead of a raw uncapped grid search keeps the run time manageable while still exploring the full realistic search space.

In [11]:
# Applicant 33 — top of the probe ranking (highest total sensitivity score)
MAX_COST_GENEROUS = 50.0   # practical "no cap": any attack > 50 is unrealistically expensive

top_probe_id = int(probe_df.iloc[0]["Applicant_ID"])
top_rejected = rejected[rejected["Applicant_ID"] == top_probe_id].iloc[0].copy()

print(f"Top probe-ranked applicant: ID={top_probe_id}")
print(f"  Approval probability : {top_rejected['approval_prob']:.4f}")
print(f"  Probe total_gain     : {probe_df.iloc[0]['total_gain']:.5f}")
print(f"  Cost cap             : {MAX_COST_GENEROUS} (generous — covers any realistic attack)")
print(f"  Caps                 : SEARCH_CAPS_FULL\n")

top_cost, top_row = find_optimal_attack_cost_capped(
    top_rejected,
    feature_names,
    max_cost=MAX_COST_GENEROUS,
    caps=SEARCH_CAPS_FULL,
)

if top_row is None:
    print(f"No flip found for Applicant {top_probe_id} under cost cap {MAX_COST_GENEROUS}.")
    top_cost = float("inf")
else:
    print(f"\nMinimum cost to flip Applicant {top_probe_id}: {top_cost:.2f}")
    top_label, top_prob = predict_single(top_row)
    print(f"Prediction after attack: {'APPROVED' if top_label == 1 else 'REJECTED'} "
          f"| P(approve)={top_prob:.4f}")

    top_comparison = pd.DataFrame({
        "Feature":  feature_names,
        "Original": top_rejected[feature_names].values.astype(float),
        "Modified": top_row,
        "Change":   top_row - top_rejected[feature_names].values.astype(float),
    })
    print("\nOriginal vs attacked values:")
    print(top_comparison.to_string(index=False))

Top probe-ranked applicant: ID=33
  Approval probability : 0.1739
  Probe total_gain     : 0.16308
  Cost cap             : 50.0 (generous — covers any realistic attack)
  Caps                 : SEARCH_CAPS_FULL

  Grid: Loan=13 Income=161 Co=101 Term=109 total=23,041,837
  Combinations checked: 23,041,837 | within cap: 8,875
No flip found for Applicant 33 under cost cap 50.0.


### Fallback: Attack the Highest Approval-Probability Applicant

Since the probe's top candidate (Applicant 33) cannot be flipped due to the immutable `Past_Bankruptcies = 1` constraint, we fall back to the next natural heuristic: **highest raw approval probability** (Applicant 13, p = 0.2457, no bankruptcy).

This attack establishes the **baseline minimum cost**, which we then use to justify the cost cap in the full search.

In [12]:
# Applicant 13 — highest approval probability among all rejected applicants
baseline_id = int(rejected.iloc[0]["Applicant_ID"])   # sorted by approval_prob desc
baseline_row = rejected.iloc[0].copy()

print(f"Baseline applicant: ID={baseline_id}")
print(f"  Approval probability : {baseline_row['approval_prob']:.4f}")
print(f"  Past_Bankruptcies    : {int(baseline_row['Past_Bankruptcies'])}")
print(f"  Cost cap             : {MAX_COST_GENEROUS} (generous)\n")

baseline_cost, baseline_modified = find_optimal_attack_cost_capped(
    baseline_row,
    feature_names,
    max_cost=MAX_COST_GENEROUS,
    caps=SEARCH_CAPS_FULL,
)

if baseline_modified is None:
    print(f"No flip found for Applicant {baseline_id} under cost cap {MAX_COST_GENEROUS}.")
else:
    print(f"\nMinimum cost to flip Applicant {baseline_id}: {baseline_cost:.2f}")
    lbl, prob = predict_single(baseline_modified)
    print(f"Prediction: {'APPROVED' if lbl == 1 else 'REJECTED'} | P(approve)={prob:.4f}")

    baseline_comparison = pd.DataFrame({
        "Feature":  feature_names,
        "Original": baseline_row[feature_names].values.astype(float),
        "Modified": baseline_modified,
        "Change":   baseline_modified - baseline_row[feature_names].values.astype(float),
    })
    print("\nOriginal vs attacked values:")
    print(baseline_comparison.to_string(index=False))

    print(f"\n→ Baseline cost = {baseline_cost:.2f}")
    print(f"→ MAX_COST = {MAX_COST} (rounded up from baseline for the full search)")

Baseline applicant: ID=13
  Approval probability : 0.2457
  Past_Bankruptcies    : 0
  Cost cap             : 50.0 (generous)

  Grid: Loan=68 Income=161 Co=101 Term=109 total=120,526,532
  Combinations checked: 120,526,532 | within cap: 13,572

Minimum cost to flip Applicant 13: 14.00
Prediction: APPROVED | P(approve)=0.7162

Original vs attacked values:
              Feature  Original  Modified  Change
                  Age      46.0      46.0     0.0
Credit_History_Length      10.0      10.0     0.0
Requested_Loan_Amount   33064.0   33064.0     0.0
      Declared_Income   40663.0   40663.0     0.0
  Co_Applicant_Income       0.0       0.0     0.0
     Loan_Term_Months      12.0      19.0     7.0
    Past_Bankruptcies       0.0       0.0     0.0

→ Baseline cost = 14.00
→ MAX_COST = 20 (rounded up from baseline for the full search)


In [13]:
# Search functions (find_optimal_attack_cost_capped, run_all_rejected_with_cost_cap)
# are defined at cell 14 — no duplicate needed here.

## 8. Cost-Cap Search on All Rejected Applicants (Probe-Ranked Order)

`MAX_COST` is derived from the baseline attack above: the cheapest observed flip costs **14**, and we apply a ~43% safety margin → `MAX_COST = 20`.

This is a principled, data-driven threshold: tight enough to prune the vast majority of the search space, yet large enough to guarantee we cannot miss a cheaper flip than the one already found.

**How the probe improves this search:**
- Applicants are processed in **probe-sensitivity order** (highest `total_gain` first), so the most attackable candidates are evaluated first.
- All 4 mutable features are always searched. The probe is **not** used to eliminate features — a zero 1-cost-unit gain does not rule out a useful larger jump (the winning +7 months on Loan_Term for Applicant 13 is a prime example).

In [14]:
# Re-order rejected applicants by total_gain (probe ranking) instead of raw approval_prob.
# Probe ranking is used ONLY for ordering — no applicant is skipped.
probe_order = probe_df["Applicant_ID"].tolist()
rejected_probe_sorted = rejected.set_index("Applicant_ID").loc[probe_order].reset_index()

print(f"Running cost-cap search on all {len(rejected)} rejected applicants")
print(f"  Order: highest probe sensitivity first")
print(f"  All 4 mutable features searched for every applicant (no feature pruning)")
print(f"  MAX_COST = {MAX_COST:.2f} | caps = SEARCH_CAPS_BUDGET")
print(f"  Only combinations with total cost ≤ {MAX_COST:.2f} are queried — provably complete.\n")

capped_results = run_all_rejected_with_cost_cap(
    rejected_df=rejected_probe_sorted,
    max_cost=MAX_COST,
    caps=SEARCH_CAPS_BUDGET,
)

print(f"\n{'='*60}")
print(f"Feasible attacks under cost cap {MAX_COST:.2f}: {len(capped_results)} / {len(rejected)}")

if len(capped_results) > 0:
    print(f"Best cost found: {capped_results[0][1]:.2f} (Applicant {capped_results[0][0]})")
else:
    print("No successful attack found under the chosen cost cap.")

Running cost-cap search on all 50 rejected applicants
  Order: highest probe sensitivity first
  All 4 mutable features searched for every applicant (no feature pruning)
  MAX_COST = 20.00 | caps = SEARCH_CAPS_BUDGET
  Only combinations with total cost ≤ 20.00 are queried — provably complete.


=== Applicant 33 (rank 1, p=0.1739) ===
  Grid: Loan=13 Income=5 Co=7 Term=11 total=5,005
  Combinations checked: 5,005 | within cap: 581
  ✗ No flip found under cost cap

=== Applicant 28 (rank 2, p=0.0615) ===
  Grid: Loan=21 Income=5 Co=7 Term=11 total=8,085
  Combinations checked: 8,085 | within cap: 616
  ✗ No flip found under cost cap

=== Applicant 13 (rank 3, p=0.2457) ===
  Grid: Loan=21 Income=5 Co=7 Term=11 total=8,085
  Combinations checked: 8,085 | within cap: 616
  ✓ Best cost: 14.00

=== Applicant 20 (rank 4, p=0.0741) ===
  Grid: Loan=21 Income=5 Co=7 Term=11 total=8,085
  Combinations checked: 8,085 | within cap: 616
  ✗ No flip found under cost cap

=== Applicant 36 (rank 5, p=

## 8b. Inspecting the Tied Applicants (13 and 48)

Both Applicants 13 and 48 achieved a best cost of **14.00**. Below we show which features were changed for each to understand the attack and decide which applicant to present as the final result.

In [15]:
TIED_IDS = [13, 48]

for app_id in TIED_IDS:
    match = [(aid, cost, mod_row, orig_row)
             for aid, cost, mod_row, orig_row in capped_results
             if aid == app_id]
    if not match:
        print(f"Applicant {app_id}: not in capped_results\n")
        continue

    _, cost, mod_row, orig_row = match[0]
    comparison = pd.DataFrame({
        "Feature":  feature_names,
        "Original": orig_row[feature_names].values.astype(float),
        "Modified": mod_row,
        "Change":   mod_row - orig_row[feature_names].values.astype(float),
    })
    changed = comparison[comparison["Change"] != 0]

    print(f"=== Applicant {app_id} | Best cost: {cost:.2f} ===")
    print(comparison.to_string(index=False))
    print(f"\nChanged features only:")
    print(changed.to_string(index=False))
    print()


=== Applicant 13 | Best cost: 14.00 ===
              Feature  Original  Modified  Change
                  Age      46.0      46.0     0.0
Credit_History_Length      10.0      10.0     0.0
Requested_Loan_Amount   33064.0   33064.0     0.0
      Declared_Income   40663.0   40663.0     0.0
  Co_Applicant_Income       0.0       0.0     0.0
     Loan_Term_Months      12.0      19.0     7.0
    Past_Bankruptcies       0.0       0.0     0.0

Changed features only:
         Feature  Original  Modified  Change
Loan_Term_Months      12.0      19.0     7.0

=== Applicant 48 | Best cost: 14.00 ===
              Feature  Original  Modified  Change
                  Age      56.0      56.0     0.0
Credit_History_Length       3.0       3.0     0.0
Requested_Loan_Amount   41913.0   41913.0     0.0
      Declared_Income   34858.0   34858.0     0.0
  Co_Applicant_Income   40000.0   40000.0     0.0
     Loan_Term_Months      12.0      19.0     7.0
    Past_Bankruptcies       0.0       0.0     0.0

Chan

### Minimum Loan Term Increase to Flip

Since only `Loan_Term_Months` changed for both applicants, we test each monthly increment one at a time to find the exact minimum that flips the prediction.

In [16]:
term_idx = feature_names.index("Loan_Term_Months")

# Stores the best fractional result across tied applicants
best_frac_cost = float("inf")
best_frac_row  = None
best_frac_orig = None
best_frac_id   = None

for app_id in TIED_IDS:
    row = rejected[rejected["Applicant_ID"] == app_id].iloc[0].copy()
    orig = row[feature_names].values.astype(float)
    term_orig = orig[term_idx]

    orig_series = row[feature_names].copy()
    print(f"Applicant {app_id} — original Loan_Term_Months = {term_orig:.0f}")

    # Step 1: find the first integer delta that flips
    flip_integer = None
    print(f"  Integer steps:")
    print(f"  {'Delta':>8}  {'Term':>8}  {'Cost':>6}  {'Result':>12}  {'P(approve)':>12}")
    for delta in range(1, 20):
        candidate = orig.copy()
        candidate[term_idx] = term_orig + delta
        label, prob = predict_single(candidate)
        cand_series = pd.Series(candidate, index=feature_names)
        with redirect_stdout(io.StringIO()):
            cost = calculate_attack_cost(pd.DataFrame([orig_series]), pd.DataFrame([cand_series]))
        verdict = "APPROVED ✓" if label == 1 else "rejected"
        print(f"  {delta:>+8}  {term_orig+delta:>8.0f}  {cost:>6.1f}  {verdict:>12}  {prob:>12.4f}")
        if label == 1:
            flip_integer = delta
            break

    if flip_integer is None:
        print("  No flip found within +20 months\n")
        continue

    # Step 2: binary-search for the minimum fractional delta
    lo, hi = float(flip_integer - 1), float(flip_integer)
    for _ in range(40):
        mid = (lo + hi) / 2
        candidate = orig.copy()
        candidate[term_idx] = term_orig + mid
        label, _ = predict_single(candidate)
        if label == 1:
            hi = mid
        else:
            lo = mid

    min_delta = hi + 1e-9
    min_term  = term_orig + min_delta

    frac_row = orig.copy()
    frac_row[term_idx] = min_term
    lbl, prob = predict_single(frac_row)

    # Use the official Cost.txt function for the final reported cost
    frac_series = pd.Series(frac_row, index=feature_names)
    with redirect_stdout(io.StringIO()):
        min_cost = calculate_attack_cost(
            pd.DataFrame([orig_series]), pd.DataFrame([frac_series])
        )

    print(f"\n  Fractional binary search result:")
    print(f"  Minimum delta : +{min_delta:.6f} months")
    print(f"  New term      : {min_term:.6f} months")
    print(f"  Cost (Cost.txt): {min_cost:.4f}")
    print(f"  Prediction    : {'APPROVED ✓' if lbl == 1 else 'REJECTED'} | P(approve)={prob:.4f}\n")

    # Keep the best fractional result (lowest cost, prefer applicant 13 on tie)
    if min_cost < best_frac_cost or (min_cost == best_frac_cost and app_id == 13):
        best_frac_cost = min_cost
        best_frac_row  = frac_row
        best_frac_orig = row
        best_frac_id   = app_id

print(f"Best fractional result: Applicant {best_frac_id}, cost = {best_frac_cost:.4f}")


Applicant 13 — original Loan_Term_Months = 12
  Integer steps:
     Delta      Term    Cost        Result    P(approve)
        +1        13     2.0      rejected        0.2457
        +2        14     4.0      rejected        0.2457
        +3        15     6.0      rejected        0.2457
        +4        16     8.0      rejected        0.2457
        +5        17    10.0      rejected        0.2457
        +6        18    12.0      rejected        0.2457
        +7        19    14.0    APPROVED ✓        0.7162

  Fractional binary search result:
  Minimum delta : +6.000001 months
  New term      : 18.000001 months
  Cost (Cost.txt): 12.0000
  Prediction    : APPROVED ✓ | P(approve)=0.7162

Applicant 48 — original Loan_Term_Months = 12
  Integer steps:
     Delta      Term    Cost        Result    P(approve)
        +1        13     2.0      rejected        0.1680
        +2        14     4.0      rejected        0.1680
        +3        15     6.0      rejected        0.1680
       

## 9. Demonstration of Final Results

In [17]:
# Section separator

In [18]:
print('=' * 65)
print('FINAL RESULTS SUMMARY')
print('=' * 65)

# --- Step 1: Probe top candidate ---
print(f'\n[1] Probe top candidate: Applicant {top_probe_id} (highest sensitivity score)')
print(f'    Baseline approval prob : {rejected[rejected["Applicant_ID"]==top_probe_id].iloc[0]["approval_prob"]:.4f}')
print(f'    Attack result          : No flip found (Past_Bankruptcies=1 acts as immutable ceiling)')

# --- Step 2: Baseline attack ---
print(f'\n[2] Baseline attack: Applicant {baseline_id} (highest approval probability)')
print(f'    Baseline approval prob : {baseline_row["approval_prob"]:.4f}')
print(f'    Minimum cost found     : {baseline_cost:.2f}')
print(f'    → MAX_COST set to {MAX_COST} (rounded up from {baseline_cost:.0f})')

# --- Step 3: Full capped search ---
winner_id, winner_cost, winner_row, winner_orig = capped_results[0]
print(f'\n[3] Cost-capped search over all 50 rejected applicants (MAX_COST={MAX_COST})')
print(f'    Successful flips found : {len(capped_results)} / 50')
print(f'    Best integer cost      : {winner_cost:.2f} (Applicants {[aid for aid,c,_,_ in capped_results if c==winner_cost]})')

# --- Step 4: Tie-breaking ---
tied = [(aid, c) for aid, c, _, _ in capped_results if c == winner_cost]
if len(tied) > 1:
    print(f'\n[4] Tie-breaking: {len(tied)} applicants share cost {winner_cost:.2f}')
    for tid, tc in tied:
        p = rejected[rejected['Applicant_ID']==tid].iloc[0]['approval_prob']
        print(f'    Applicant {tid}: approval_prob={p:.4f}')
    print(f'    → Chose Applicant {winner_id} (highest baseline approval probability)')

# --- Step 5: Fractional refinement ---
print(f'\n[5] Fractional term refinement (binary search on Loan_Term_Months)')
print(f'    Best fractional cost   : {best_frac_cost:.4f} (Applicant {best_frac_id})')
print(f'    Improvement over integer result: {winner_cost:.2f} → {best_frac_cost:.4f}')

# --- Step 6: Final adversarial example ---
frac_label, frac_prob = predict_single(best_frac_row)
print(f'\n[6] Final Adversarial Applicant: ID = {best_frac_id}')
print(f'    Total modification cost : {best_frac_cost:.4f}')
print(f'    Post-attack prediction  : {"APPROVED" if frac_label == 1 else "REJECTED"}')
print(f'    Post-attack P(approve)  : {frac_prob:.4f}')

final_comparison = pd.DataFrame({
    'Feature':  feature_names,
    'Original': best_frac_orig[feature_names].values.astype(float),
    'Modified': best_frac_row,
    'Change':   best_frac_row - best_frac_orig[feature_names].values.astype(float),
})
print(f'\nOriginal vs Adversarial values:')
print(final_comparison.to_string(index=False))
changed = final_comparison[final_comparison['Change'] != 0]
print(f'\nModified features only:')
print(changed.to_string(index=False))
print('\n' + '=' * 65)


FINAL RESULTS SUMMARY

[1] Probe top candidate: Applicant 33 (highest sensitivity score)
    Baseline approval prob : 0.1739
    Attack result          : No flip found (Past_Bankruptcies=1 acts as immutable ceiling)

[2] Baseline attack: Applicant 13 (highest approval probability)
    Baseline approval prob : 0.2457
    Minimum cost found     : 14.00
    → MAX_COST set to 20 (rounded up from 14)

[3] Cost-capped search over all 50 rejected applicants (MAX_COST=20)
    Successful flips found : 2 / 50
    Best integer cost      : 14.00 (Applicants [13, 48])

[4] Tie-breaking: 2 applicants share cost 14.00
    Applicant 13: approval_prob=0.2457
    Applicant 48: approval_prob=0.1680
    → Chose Applicant 13 (highest baseline approval probability)

[5] Fractional term refinement (binary search on Loan_Term_Months)
    Best fractional cost   : 12.0000 (Applicant 13)
    Improvement over integer result: 14.00 → 12.0000

[6] Final Adversarial Applicant: ID = 13
    Total modification cost : 1